includes/class-wc-product-purchases-helper.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+/**
+ * WC_Product_Purchases_Helper class.
+ *
+ * @class     WC_Product_Purchases_Helper
+ * @version   3.3.0
+ * @package   WooCommerce/Classes
+ * @category  Class
+ * @author    WooThemes
+ */
+class WC_Product_Purchases_Helper {
+
+	/**
+	 * Hook in methods.
+	 */
+	public static function init() {
+		add_action( 'woocommerce_delete_order', array( __CLASS__, 'handle_delete_order' ), 10, 1 );
+		add_action( 'woocommerce_order_object_updated_props', array( __CLASS__, 'handle_product_purchase' ), 10, 2 );
+	}
+
+	/**
+	 * Handler for deleting an order, keep the purchased products table in sync.
+	 *
+	 * @param int $order_id
+	 */
+	public static function handle_delete_order( $order_id ) {
+		global $wpdb;
+
+		$wpdb->delete( $wpdb->prefix . 'woocommerce_product_purchases', array( 'order_id' => $order_id ), array( '%d' ) );
+	}
+
+	/**
+	 * Handler for saving an order, keep the purchased products table in sync.
+	 *
+	 * @param WC_Order $order
+	 * @param array    $updated_props
+	 */
+	public static function handle_product_purchase( $order, $updated_props ) {
+		global $wpdb;
+
+		// Every order is marked as paid at most once.
+		// Only when this happens we're updating the purchased products table.
+		if ( ! in_array( 'date_paid', $updated_props ) || empty( $order->get_date_paid() ) ) {
```

Comment:
This seemed to work fine on manual testings, but is causing some UT failures. The orders there have `needs_processing` returning true, so not sure if we need to alter this conditional or modify how the orders are created in the tests helper framework.